PUBDEV-6089: AutoML expose eventLog on Py+R clients

seb-h2o opened the pull request:
https://0xdata.atlassian.net/browse/PUBDEV-6089

- renamed automl UserFeedback to EventLog as old name was confusing.
- expose new EventLog to Py and R clients.
- extended with some key/values for post analysis: cf. https://0xdata.atlassian.net/browse/PUBDEV-5316
- added the live event log display when polling automl: https://0xdata.atlassian.net/browse/PUBDEV-6090


examples of rendering of "live event log" during automl training:

R client:
```
[1] "verbosity disabled"
  |======================================================================| 100%
[1] "verbosity level = info"
  |                                                                      |   0%
15:23:59.877: Project: r_test_train_verbosity_info
15:23:59.877: AutoML job created: 2019.05.30 15:23:59.877
15:23:59.877: Build control seed: 1234
15:23:59.877: training frame: Frame key: automl_training_RTMP_sid_8731_56    cols: 9    rows: 380  chunks: 1    size: 5426  checksum: -719866744755860782
15:23:59.877: validation frame: NULL
15:23:59.877: leaderboard frame: NULL
15:23:59.877: response column: CAPSULE
15:23:59.877: fold column: null
15:23:59.877: weights column: null
15:23:59.878: Setting stopping tolerance adaptively based on the training frame: 0.05
15:23:59.878: AutoML build started: 2019.05.30 15:23:59.878
15:23:59.878: XGBoost_1_AutoML_20190530_152359 started
15:24:00.882: XGBoost_1_AutoML_20190530_152359 complete
15:24:00.883: New leader: XGBoost_1_AutoML_20190530_152359, auc: 0.7981918171086348
  |===================================================================   |  96%
15:24:01.887: XGBoost_2_AutoML_20190530_152359 complete
15:24:01.889: StackedEnsemble build using top model from each algorithm type started
15:24:02.890: StackedEnsemble build using top model from each algorithm type complete
15:24:02.892: StackedEnsemble build using all AutoML models started
15:24:03.893: StackedEnsemble build using all AutoML models complete
15:24:03.894: AutoML build stopped: 2019.05.30 15:24:03.894
  |======================================================================| 100%
[1] "verbosity level = warn"
  |                                                                      |   0%
15:24:09.549: Stopping tolerance set by the user is < 70% of the recommended default of 0.05, so models may take a long  |======================================================================| 100%
```

Python client:
```
verbosity off
AutoML progress: |████████████████████████████████████████████████████████████████████████████| 100%
verbosity info
AutoML progress: |                                                                            |   0%
19:45:23.571: Project: test_train_verbosity
19:45:23.571: AutoML job created: 2019.05.30 19:45:23.571
19:45:23.571: Build control seed: 1234
19:45:23.571: training frame: Frame key: automl_training_py_56_sid_82b9    cols: 9    rows: 380  chunks: 1    size: 5538  checksum: -719866744755860782
19:45:23.571: validation frame: NULL
19:45:23.571: leaderboard frame: NULL
19:45:23.571: response column: CAPSULE
19:45:23.571: fold column: null
19:45:23.571: weights column: null
19:45:23.571: Stopping tolerance set by the user: 0.01
19:45:23.572: Stopping tolerance set by the user is < 70% of the recommended default of 0.05, so models may take a long time to converge or may not converge at all.
19:45:23.573: AutoML build started: 2019.05.30 19:45:23.573
19:45:23.573: XGBoost_1_AutoML_20190530_194523 started

AutoML progress: |█▌                                                                          |   2%
19:45:25.581: XGBoost_1_AutoML_20190530_194523 complete
19:45:25.582: New leader: XGBoost_1_AutoML_20190530_194523, auc: 0.8076070369410613
19:45:25.583: XGBoost_2_AutoML_20190530_194523 started

AutoML progress: |██▎                                                                         |   3%
19:45:26.584: XGBoost_2_AutoML_20190530_194523 complete
19:45:26.588: StackedEnsemble build using top model from each algorithm type started

AutoML progress: |█████████████▋                                                              |  18%
19:45:27.592: StackedEnsemble build using top model from each algorithm type complete
19:45:27.593: StackedEnsemble build using all AutoML models started

AutoML progress: |████████████████████████████████████████████████████████████████████████████| 100%

19:45:28.598: StackedEnsemble build using all AutoML models complete
19:45:28.600: AutoML build stopped: 2019.05.30 19:45:28.600
19:45:28.600: AutoML build done: built 4 models

verbosity warn
AutoML progress: |                                                                            |   0%
19:45:29.497: Stopping tolerance set by the user is < 70% of the recommended default of 0.05, so models may take a long time to converge or may not converge at all.

AutoML progress: |████████████████████████████████████████████████████████████████████████████| 100%
```

seb-h2o commented on h2o-automl/src/main/java/ai/h2o/automl/AutoML.java:
```diff
@@ -1,6 +1,6 @@
 package ai.h2o.automl;
```
mainly renaming in this class: UserFeedback -> EventLog

seb-h2o commented on h2o-automl/src/main/java/ai/h2o/automl/EventLog.java:
```diff
@@ -0,0 +1,98 @@
+package ai.h2o.automl;
```
renamed from UserFeedback: no fundamental changes

seb-h2o commented on h2o-automl/src/main/java/ai/h2o/automl/EventLogEntry.java:
```diff
@@ -0,0 +1,192 @@
+package ai.h2o.automl;
```
– renamed from UserFeedbackEvent.
– added support for event metadata (1 per event): cf. setNamedValue()

seb-h2o commented on h2o-automl/src/main/java/ai/h2o/automl/FrameMetadata.java:
```diff
@@ -1,6 +1,6 @@
 package ai.h2o.automl;
```
renaming Userfeedback -> EventLog

seb-h2o commented on h2o-automl/src/main/java/ai/h2o/automl/Leaderboard.java:
```diff
@@ -102,10 +102,9 @@
   private boolean have_set_sort_metric = false;
```
renaming Userfeedback -> EventLog

seb-h2o commented on h2o-automl/src/main/java/water/automl/api/schemas3/AutoMLV99.java:
```diff
@@ -1,18 +1,29 @@
+      super(key);
+    }
+  }
+
```
moved this schema class out of AutoML to group it with other schema definitions

deil87 replied:
@seb-h2o  where was it before I wonder? :)  is it related to eventLog stuff?

seb-h2o replied:
na, it was in AutoML

seb-h2o commented on h2o-automl/src/main/java/water/automl/api/schemas3/AutoMLV99.java:
```diff
@@ -1,18 +1,29 @@
 package water.automl.api.schemas3;
```
mainly renamings

seb-h2o commented on h2o-automl/src/main/java/water/automl/api/schemas3/EventLogEntryV99.java:
```diff
@@ -0,0 +1,48 @@
+package water.automl.api.schemas3;
```
renamed from UserfeedbackEventV99 + exposed 2 new fields: name, value

seb-h2o commented on h2o-automl/src/main/java/water/automl/api/schemas3/EventLogV99.java:
```diff
@@ -0,0 +1,30 @@
+package water.automl.api.schemas3;
```
renamed from UserFeedbackV99

deil87 replied:
@seb-h2o when the file name gets changed it is a tough task to check what was really changed :)

seb-h2o replied:
you can check the old ones, but that's why I write additional changes on top file: this one is pure renaming

deil87 replied:
@seb-h2o I got your intents.. just wanted to express my compassion :)

seb-h2o commented on h2o-automl/src/test/java/ai/h2o/automl/LeaderboardTest.java:
```diff
@@ -5,12 +5,10 @@
 import org.junit.Assert;
```
renamings

seb-h2o commented on h2o-bindings/bin/bindings.py:
```diff
@@ -345,7 +345,7 @@ def schemas_map(add_generics=False):
     m = {}
```
renamings

seb-h2o commented on h2o-py/h2o/automl/autoh2o.py:
```diff
@@ -303,6 +327,8 @@ def train(self, x = None, y = None, training_frame = None, fold_column = None,
             This is optional, but when provided, it is also recommended to disable cross validation 
             by setting `nfolds=0` and to provide a leaderboard frame for scoring purposes.
+        :param verbosity: Verbosity of the backend messages printed during training.
+            Available options are 'debug', 'info' or 'warn'. Defaults to None (disable live log).
```
@ledell I've been hesitating to default verbosity param to 'warn' level, same for R API. Any preference?

seb-h2o commented on h2o-automl/src/main/java/ai/h2o/automl/AutoML.java:
```diff
@@ -351,11 +342,13 @@ public void stop() {
+    eventLog.info(Stage.Workflow, "AutoML build stopped: " + EventLogEntry.dateTimeFormat.format(runCountdown.stop_time()))
+            .setNamedValue("stop_epoch", runCountdown.stop_time(), EventLogEntry.epochFormat);
+    eventLog.info(Stage.Workflow, "AutoML build done: built " + modelCount + " models")
+            .setNamedValue("model_count", modelCount);
```
examples of exposed meta-info for post-analysis

seb-h2o commented on h2o-py/h2o/automl/autoh2o.py:
```diff
@@ -470,35 +500,92 @@ def download_mojo(self, path=".", get_genmodel_jar=False, genmodel_name=""):
+                    state['last_events_nrows'] = events.nrows
+            state['last_job_progress'] = job.progress
+        except Exception as e:
+            print("Failed polling AutoML progress log: {}".format(e))
```
handles the "live" print during automl training

seb-h2o commented on h2o-r/h2o-package/R/automl.R:
```diff
@@ -54,6 +54,8 @@
 #'        "mean_per_class_error" for multinomial classification, and "mean_residual_deviance" for regression.
 #' @param export_checkpoints_dir (Optional) Path to a directory where every model will be stored in binary form.
+#' @param verbosity Verbosity of the backend messages printed during training; Optional.
+#'        Must be one of "debug", "info", "warn". Defaults to NULL (disable live log).
```
cf. concern about defaults in Python client

seb-h2o commented:
examples of rendering of "live event log" during automl training:

R client:
```
[1] "verbosity disabled"
  |======================================================================| 100%
[1] "verbosity level = info"
  |                                                                      |   0%
15:23:59.877: Project: r_test_train_verbosity_info
15:23:59.877: AutoML job created: 2019.05.30 15:23:59.877
15:23:59.877: Build control seed: 1234
15:23:59.877: training frame: Frame key: automl_training_RTMP_sid_8731_56    cols: 9    rows: 380  chunks: 1    size: 5426  checksum: -719866744755860782
15:23:59.877: validation frame: NULL
15:23:59.877: leaderboard frame: NULL
15:23:59.877: response column: CAPSULE
15:23:59.877: fold column: null
15:23:59.877: weights column: null
15:23:59.878: Setting stopping tolerance adaptively based on the training frame: 0.05
15:23:59.878: AutoML build started: 2019.05.30 15:23:59.878
15:23:59.878: XGBoost_1_AutoML_20190530_152359 started
15:24:00.882: XGBoost_1_AutoML_20190530_152359 complete
15:24:00.883: New leader: XGBoost_1_AutoML_20190530_152359, auc: 0.7981918171086348
  |===================================================================   |  96%
15:24:01.887: XGBoost_2_AutoML_20190530_152359 complete
15:24:01.889: StackedEnsemble build using top model from each algorithm type started
15:24:02.890: StackedEnsemble build using top model from each algorithm type complete
15:24:02.892: StackedEnsemble build using all AutoML models started
15:24:03.893: StackedEnsemble build using all AutoML models complete
15:24:03.894: AutoML build stopped: 2019.05.30 15:24:03.894
  |======================================================================| 100%
[1] "verbosity level = warn"
  |                                                                      |   0%
15:24:09.549: Stopping tolerance set by the user is < 70% of the recommended default of 0.05, so models may take a long  |======================================================================| 100%
```

Python client:
```
verbosity off
AutoML progress: |████████████████████████████████████████████████████████████████████████████| 100%
verbosity info
AutoML progress: |                                                                            |   0%
15:16:35.344: Project: test_train_verbosity
15:16:35.345: AutoML job created: 2019.05.30 15:16:35.344
15:16:35.345: Build control seed: 1234
15:16:35.345: training frame: Frame key: automl_training_py_4_sid_8d03    cols: 9    rows: 380  chunks: 1    size: 5538  checksum: -719866744755860782
15:16:35.345: validation frame: NULL
15:16:35.345: leaderboard frame: NULL
15:16:35.345: response column: CAPSULE
15:16:35.345: fold column: null
15:16:35.345: weights column: null
15:16:35.345: Stopping tolerance set by the user: 0.01
15:16:35.345: Stopping tolerance set by the user is < 70% of the recommended default of 0.05, so models may take a long time to converge or may not converge at all.
15:16:35.346: AutoML build started: 2019.05.30 15:16:35.346
15:16:35.347: XGBoost_1_AutoML_20190530_151635 started

AutoML progress: |█▋                                                                          |   2%
15:16:37.356: XGBoost_1_AutoML_20190530_151635 complete
15:16:37.358: XGBoost_2_AutoML_20190530_151635 started

AutoML progress: |██▍                                                                         |   3%
15:16:38.363: XGBoost_2_AutoML_20190530_151635 complete
15:16:38.366: StackedEnsemble build using top model from each algorithm type started

AutoML progress: |█████████████▌                                                              |  18%
15:16:39.371: StackedEnsemble build using top model from each algorithm type complete
15:16:39.373: StackedEnsemble build using all AutoML models started

AutoML progress: |████████████████████████████████████████████████████████████████████████████| 100%
verbosity warn
AutoML progress: |                                                                            |   0%
15:16:41.61: Stopping tolerance set by the user is < 70% of the recommended default of 0.05, so models may take a long time to converge or may not converge at all.

AutoML progress: |████████████████████████████████████████████████████████████████████████████| 100%
```

seb-h2o commented:
@ledell, @deil87 : this is complete, please have a look: most of the changes are Py + R clients

deil87 commented:
@seb-h2o maybe it is just not a complete print out for Python case but I don't see ` AutoML build stopped` message there.

deil87 commented on h2o-automl/src/test/java/ai/h2o/automl/EventLogTest.java:
```diff
@@ -0,0 +1,58 @@
+            eventLog.info(Stage.ModelTraining, "this doesn't mean what it looks").setNamedValue("bar", 777);
+
+            TwoDimTable events = eventLog.toTwoDimTable("Test header");
+            Assert.assertEquals(6, events.getColDim());
```
@seb-h2o maybe it is unnecessary to check number of columns as you are checking it with headers afterwards.

seb-h2o replied:
true, but I started with dims indeed :) can't hurt

deil87 commented on h2o-automl/src/test/java/ai/h2o/automl/EventLogTest.java:
```diff
@@ -0,0 +1,58 @@
+                                    events.getColHeaders());
+
+            DateTime now = DateTime.now();
+            DateTime entry_time = new DateTime(EventLogEntry.timeFormat.parse((String)events.get(0, 0)).getTime());
```
@seb-h2o snake style sneaked in :)

deil87 commented on h2o-r/h2o-package/R/models.R:
```diff
@@ -471,11 +481,7 @@ predict.H2OModel <- function(object, newdata, ...) {
 #' @rdname predict.H2OModel
 #' @export
 h2o.predict <- function(object, newdata, ...){
-  if(class(object) == "H2OAutoML"){
```
@seb-h2o it looks like it could be implemented via generic function (for `predict`) and dispatch concept. I'm just curious how `UseMethod` is working. It looks like it is built in function... but it is strange that name is camel case :)

seb-h2o replied:
it's doing exactly what what you said: it defines "predict" as a generic function, and the dispatching is done based on the object type.
e.g. if object is a H2OAutoML, it dispatches to predict.H2OAutoML

deil87 replied:
@seb-h2o great! good to know!

seb-h2o commented on h2o-py/h2o/automl/autoh2o.py:
```diff
@@ -275,11 +280,30 @@ def leaderboard(self):
+        return H2OFrame([]) if self._event_log is None else self._event_log
+
+    @property
+    def training_info(self):
```
@ledell feel free to come up with a better name, as this may come handy to expose more info in the future for post-analysis

ledell replied:
@seb-h2o What about something more generic, e.g. `event_log()` or `log_info()`?  Using the term "training" to me, its not immediately clear whether that refers to the training set or model training.

seb-h2o replied:
@ledell  I use `event_log` property right above for the full EventLog frame: 6 cols, n rows.
This `training_info` is just sugar on top: it extracts log entries with key-values as a dictionary for ease of use. Those key-values are mainly useful for post-analysis as we discussed: https://0xdata.atlassian.net/browse/PUBDEV-5316

seb-h2o replied:
`log_info` maybe. I'm not happy with `training_info` either: I was mainly looking for a word that  says it contains meta information about the training that can be useful for post-analysis.

seb-h2o commented on h2o-py/h2o/estimators/estimator_base.py:
```diff
@@ -246,11 +247,29 @@ def _train(self, x=None, y=None, training_frame=None, offset_column=None, fold_c
+            print("Model build is starting now...")
+        except AttributeError:  # To catch NoneType error if scoring history is not available
+            print("Scoring History is not available yet...")
+
```
moved from `job.py` as it's specific to model training jobs...
instead, `job.poll()` method takes a callback function now

ledell commented:
I'm adding some more background on what the logs look like:

R:
```
> aml <- h2o.automl(y = "Species", training_frame = as.h2o(iris), max_models = 10)
> aml@event_log
     timestamp level      stage
1 13:46:25.423  Info   Workflow
2 13:46:25.423  Info   Workflow
3 13:46:25.424  Info   Workflow
4 13:46:25.425  Info DataImport
5 13:46:25.425  Info DataImport
6 13:46:25.425  Info DataImport
                                                                                                                                     message
1                                                                                                            Project: automl_iris_sid_bf43_1
2                                                                                                AutoML job created: 2019.06.13 13:46:25.420
3                                                                                                            Build control seed: -1 (random)
4 training frame: Frame key: automl_training_iris_sid_bf43_1    cols: 5    rows: 150  chunks: 1    size: 1816  checksum: 8131244627986678876
5                                                                                                                     validation frame: NULL
6                                                                                                                    leaderboard frame: NULL
            name      value
1           <NA>       <NA>
2 creation_epoch 1560458785
3           <NA>       <NA>
4           <NA>       <NA>
5           <NA>       <NA>
6           <NA>       <NA>

[52 rows x 6 columns] 

> aml@training_info
$creation_epoch
[1] "1560458785"

$start_epoch
[1] "1560458785"

$stop_epoch
[1] "1560458799"

$model_count
[1] "12"

$duration_secs
[1] "13"
```

ledell commented:
@seb-h2o and I discussed removing `model_count` since it's redundant, but other than that, we are good to go!

seb-h2o commented:
@ledell FYI added a couple of commits:
- removed `model_count`
- added documentation about `verbosity` parameter to `automl.rst`, as well a short line mentioning the `event_log` and `training_info` properties in Python and R clients.
- reuse the same key when fetching the `event_log` "live" during training in R client  to avoid leaked keys.

ledell commented:
@seb-h2o My one last question is whether `verbosity` in Python belongs in the `H2OAutoML` init or `.train()` method in Python.  So far, `.train` is stuff only related to the data we train on, where as the init params have to do with the AutoML process itself, so I wonder if maybe `verbosity` belongs there intead?

seb-h2o commented:
@ledell regarding `velocity`, it's a relevant concern. I had to chose between:
1. putting this `verbosity` param in the constructor to keep only data-related params in `train` and keep AutoML Py API self-consistent.
2. putting it in `train` method to keep it consistent with `H2OEstimator.train` method that also has a `verbose` param.

As a programmer, I prefer #1, but instinctively I tend to think that user would prefer #2, not completely sure though.
I'm fine moving it to constructor if we consider that AutoML doesn't systematically have to align with base algorithms.

ledell commented:
@seb-h2o I'm inclined to move it to the constructor.

seb-h2o commented:
@ledell moved the `verbosity` param: tests are passing now, ready to merge to `yates` as well as matching Flow PR: https://github.com/h2oai/h2o-flow/pull/140
